aemcmc/nuts.py

```diff
@@ -0,0 +1,222 @@
+
+    transformed_and_raveled = rp_map.ravel_params(transformed_rvs.values())
+
+    # Keep track of parameters ---------------------------------------
+
+    # current_samples links random variable names to their current
+    # value in the sampling process
+    # current_samples_transforms does the same in the transformed space
+    current_samples = {
+        model.model.values_to_rvs[v].name: v.tag.test_value
+        for v in model.model.unobserved_values
+    }
+    current_samples_transformed = current_samples.copy()
+
+    def uravel_and_untrans(q):
+        return {
+            rv: (q_part, transform_var(rv, q_part, forward=False))
+            for rv, q_part in rp_map.unravel_params(q).items()
+        }
+
+    def update_current_samples(q):
+        qs_unraveled, qs_unrav_untrans = zip(*uravel_and_untrans(q).values())
+        for rv, val, val_untrans in zip(
+            sampled_rvs_to_vals, qs_unraveled, qs_unrav_untrans
+        ):
+            current_samples_transformed[rv.name] = val
+            current_samples[rv.name] = val_untrans
+
+    # The trace accumulates the values and transformed values in the
+    # sampling process.
+    trace: Dict[TensorVariable, List[np.ndarray]] = {
+        k: [] for k in tuple(current_samples.keys())
+    }
+
+    q_0 = transformed_and_raveled(
+        *tuple(current_samples[v.name] for v in rp_map.ref_params)
+    )
+    update_current_samples(q_0)
+
+    nuts_value_logps_needed = set()
+    for vv in tuple(sampled_rvs_to_vals.values()):
+        nuts_value_logps_needed.update(values_to_values_deps[vv])
+
+    nuts_logps = {vv.name: values_to_logps[vv] for vv in nuts_value_logps_needed}
+
+    def nuts_logprob(q):
+        """Compute the value of the graph's logprob.
+
+        Parameter
+        ---------
+        q
+            The vectorized version of the current parameter values.
+
+        """
+
+        # Gather all the log-probabilities that depend on the
+        # NUTS-estimated parameters
+        total_logprob = at.sum(
+            tuple(at.sum(rv_logp) for rv_logp in nuts_logps.values())
+        )
+
+        unraveled_q = rp_map.unravel_params(q)
+
+        # Now, we need to replace all references to the original value
+        # variables with their unraveled values
+        vv_to_unraveled_vals = {
+            vv: unraveled_q[rv] for rv, vv in sampled_rvs_to_vals.items()
+        }
+
+        # Don't clone sub-graphs within our input parameter `q`
+        vv_to_unraveled_vals[q] = q
+
+        memo = aesara.graph.basic.clone_get_equiv(
+            [],
+            [total_logprob],
+            copy_inputs=False,
+            copy_orphans=False,
+            memo=vv_to_unraveled_vals,
+        )
+        nuts_total_logprob = memo[total_logprob]
+
+        return nuts_total_logprob
+
+    inverse_mass_matrix = at.ones_like(rp_map.unravel_params(q_0))
+    step_size = 1e-3
+
+    nuts_step_fn = aehmc.nuts.kernel(srng, nuts_logprob, inverse_mass_matrix)
```

Comment:
We should only return the step function, and, later, create a means of combining step functions into a `Scan` sampling loop.

Reply:
I fully agree, I was originally going to do this in another PR later (and refactor the gibbs samplers as well) but we can do it here if you prefer.